File: core/src/com/inkostilation/pong/engine/IPongEngine.java
package com.inkostilation.pong.engine;

import com.inkostilation.pong.commands.AbstractRequestCommand;
import com.inkostilation.pong.exceptions.NoEngineException;

import java.io.IOException;

public interface IPongEngine<M> extends IEngine<M> {
    void sendCommand(AbstractRequestCommand<IEngine<M>, M>... command) throws IOException, NoEngineException;
    void sendFieldState(M marker) throws IOException;
    void sendPlayerRole(M marker) throws IOException;
    void sendScore(M marker) throws IOException;
    void applyInput(Direction direction, M marker) throws IOException;
    void startNewGame(M marker) throws IOException;
    void connectToGame(M marker, int index) throws IOException;
    void confirm(M Marker);
    void sendGameState(M marker) throws IOException;
    void sendGameList(M marker) throws IOException;
    void sendGameUpdate(M marker) throws IOException;
}

File: core/src/com/inkostilation/pong/engine/IPongEngine.kt
package com.inkostilation.pong.engine

import java.io.IOException

interface IPongEngine<M> : IEngine<M> {
    @Throws(IOException::class)
    fun sendFieldState(marker: M)

    @Throws(IOException::class)
    fun sendPlayerRole(marker: M)

    @Throws(IOException::class)
    fun sendScore(marker: M)

    @Throws(IOException::class)
    fun applyInput(direction: Direction?, marker: M)

    @Throws(IOException::class)
    fun startNewGame(marker: M)

    @Throws(IOException::class)
    fun connectToGame(marker: M, index: Int)

    fun confirm(Marker: M)
    @Throws(IOException::class)
    fun sendGameState(marker: M)

    @Throws(IOException::class)
    fun sendGameList(marker: M)

    @Throws(IOException::class)
    fun sendGameUpdate(marker: M)
}